processing/src/main/java/org/apache/druid/segment/StringDimensionIndexer.java

```diff
@@ -132,12 +142,14 @@ public EncodedKeyComponent<int[]> processRowValsToUnsortedEncodedKeyComponent(@N
     } else if (dimValues instanceof byte[]) {
       encodedDimensionValues =
           new int[]{dimLookup.add(emptyToNullIfNeeded(StringUtils.encodeBase64String((byte[]) dimValues)))};
+      dictionaryChanged = true;
     } else {
       encodedDimensionValues = new int[]{dimLookup.add(emptyToNullIfNeeded(dimValues))};
+      dictionaryChanged = true;
     }
 
     // If dictionary size has changed, the sorted lookup is no longer valid.
-    if (oldDictSize != dimLookup.size()) {
+    if (dictionaryChanged) {
```

Comment:
only the `add` method are called on `dimLookup` above - which naturally changes the `size` ; I'm not sure what the addition of this `boolean` adds - as it must be set to true after all `add` calls.

Actually since the `add` call may even [bail out and do nothing](https://github.com/apache/druid/blob/3c72ec8413fc066dbb56e9f02d6fbc6e3b28bbe1/processing/src/main/java/org/apache/druid/segment/DimensionDictionary.java#L164) in some cases: Theoretically this may force `sortedLookup` to become `null` in cases in which earlier it would not done that - which seems like a behaviour change...

...and because of the above I wonder how does the introduction of this boolean improves performance?

Reply:
Since the dimLookup.size() is quite expensive, as seen from the screenshot, I've tried to remove that call by implicitly checking if the `add` to `dimLookup` results in a change in the size of `dimLookup`. 
You're right about the behavioural change. To keep the same behaviour, I've pushed a commit which would now check if the result of `dimLookup.add()` with `-1` and then set the boolean `dictionaryChanged` accordingly. 
Please take a look @kgyrtkirk

Reply:
I believe the main purpose of the `dimLookup` field is to provide a common place where the different strings get an assigned number - so that if its able to recall an earlier value that's actually very important ...this change will destroy the `sortedLookup` cache even in case all calls to `add` were hits.

I think you should either remove the changes in this file - or find a different way to address this; one idea could be; to connect the `StringDimensionIndexer` class as a modification listener of `dimLookup` - so that when a real add happens the `sortedLookup` will automatically destructed:
* move the sorted dictionary caching to `StringDimensionDictionary`
* extract a protected `realAdd` from the current `add` in `DimensionDictionary` 
* implement `realAdd` to clear the cached `sorted` stuff in `StringDimensionDictionary`

again; let me invite @clintropolis to see if He have a different idea :)

Reply:
>so that if its able to recall an earlier value that's actually very important ...this change will destroy the sortedLookup cache even in case all calls to add were hits.

The lookup is only sorted when persisting a segment, so `add` shouldn't really happen after the dictionary is sorted. I would guess this is here as a safeguard.

Reply:
You are right, this invalidates the cache hits as well and changes the behaviour. I've pushed a new commit which would change the return type of `add` function from `int` to `Pair<Integer, Boolean>` where the boolean indicates whether an element was actually added to the dictionary or not. The integer however remains the same. 
Please take a look @kgyrtkirk and @clintropolis

Reply:
it's too bad `useMaxMemoryEstimates` which corresponds to `StringDimensionDictionary.computeOnHeapSize` isn't mandatory, since it seems like we could simply check if `DimensionDictionary.sizeInBytes` has changed from before and after instead of all of this pair stuff...

It is on by default for native ingestion, I wonder if we should change it to always compute it in the dictionary and just ignore the value if `useMaxMemoryEstimates` is false...